File: clients/cpp/model/Unit.cpp
#include "Unit.hpp"

Unit::Unit() { }
Unit::Unit(int playerId, int id, int health, Vec2Double position, Vec2Double size, JumpState jumpState, bool walkedRight, bool stand, bool onGround, bool onLadder, int mines, std::shared_ptr<Weapon> weapon) : playerId(playerId), id(id), health(health), position(position), size(size), jumpState(jumpState), walkedRight(walkedRight), stand(stand), onGround(onGround), onLadder(onLadder), mines(mines), weapon(weapon) { }
Unit Unit::readFrom(InputStream& stream) {
    Unit result;
    result.playerId = stream.readInt();
    result.id = stream.readInt();
    result.health = stream.readInt();
    result.position = Vec2Double::readFrom(stream);
    result.size = Vec2Double::readFrom(stream);
    result.jumpState = JumpState::readFrom(stream);
    result.walkedRight = stream.readBool();
    result.stand = stream.readBool();
    result.onGround = stream.readBool();
    result.onLadder = stream.readBool();
    result.mines = stream.readInt();
    if (stream.readBool()) {
        result.weapon = std::shared_ptr<Weapon>(new Weapon());
        *result.weapon = Weapon::readFrom(stream);
    } else {
        result.weapon = std::shared_ptr<Weapon>();
    }
    return result;
}
void Unit::writeTo(OutputStream& stream) const {
    stream.write(playerId);
    stream.write(id);
    stream.write(health);
    position.writeTo(stream);
    size.writeTo(stream);
    jumpState.writeTo(stream);
    stream.write(walkedRight);
    stream.write(stand);
    stream.write(onGround);
    stream.write(onLadder);
    stream.write(mines);
    if (weapon) {
        stream.write(false);
    } else {
        stream.write(true);
        (*weapon).writeTo(stream);
    }
}
std::string Unit::toString() const {
    return std::string("Unit") + "(" +
        std::to_string(playerId) +
        std::to_string(id) +
        std::to_string(health) +
        position.toString() +
        size.toString() +
        jumpState.toString() +
        (walkedRight ? "true" : "false") + 
        (stand ? "true" : "false") + 
        (onGround ? "true" : "false") + 
        (onLadder ? "true" : "false") + 
        std::to_string(mines) +
        "TODO" + 
        ")";
}

File: clients/ruby/model/unit.rb
require_relative 'vec2_double'
require_relative 'vec2_double'
require_relative 'jump_state'
require_relative 'weapon'
class Unit
    attr_accessor :player_id
    attr_accessor :id
    attr_accessor :health
    attr_accessor :position
    attr_accessor :size
    attr_accessor :jump_state
    attr_accessor :walked_right
    attr_accessor :stand
    attr_accessor :on_ground
    attr_accessor :on_ladder
    attr_accessor :mines
    attr_accessor :weapon
    def initialize(player_id, id, health, position, size, jump_state, walked_right, stand, on_ground, on_ladder, mines, weapon)
        @player_id = player_id
        @id = id
        @health = health
        @position = position
        @size = size
        @jump_state = jump_state
        @walked_right = walked_right
        @stand = stand
        @on_ground = on_ground
        @on_ladder = on_ladder
        @mines = mines
        @weapon = weapon
    end
    def self.read_from(stream)
        player_id = stream.read_int()
        id = stream.read_int()
        health = stream.read_int()
        position = Vec2Double.read_from(stream)
        size = Vec2Double.read_from(stream)
        jump_state = JumpState.read_from(stream)
        walked_right = stream.read_bool()
        stand = stream.read_bool()
        on_ground = stream.read_bool()
        on_ladder = stream.read_bool()
        mines = stream.read_int()
        if stream.read_bool()
            weapon = Weapon.read_from(stream)
        else
            weapon = nil
        end
        Unit.new(player_id, id, health, position, size, jump_state, walked_right, stand, on_ground, on_ladder, mines, weapon)
    end
    def write_to(stream)
        stream.write_int(@player_id)
        stream.write_int(@id)
        stream.write_int(@health)
        @position.write_to(stream)
        @size.write_to(stream)
        @jump_state.write_to(stream)
        stream.write_bool(@walked_right)
        stream.write_bool(@stand)
        stream.write_bool(@on_ground)
        stream.write_bool(@on_ladder)
        stream.write_int(@mines)
        if @weapon.nil?
            stream.write_bool(false)
        else
            stream.write_bool(true)
            @weapon.write_to(stream)
        end
    end
end
